src/Recording.ts

```diff
@@ -172,53 +199,148 @@ export default class Recording {
     return event;
   }
 
-  private emit(event: unknown) {
-    // Check here if we should record instead of requiring each
-    // possible hook to check it.
-    // This is also checked in recorder.record() to prevent
-    // unnecessary event object creation. Checking this inside hooks,
-    // (http, sqlite, pg, mysql, ...) will save some CPU cycles but
-    // will complicate their code.
-    if (!shouldRecord()) return;
-    assert(this.stream);
-    this.stream.emit(event);
-  }
-
   private eventUpdates: Record<number, AppMap.Event> = {};
 
   fixup(event: AppMap.Event) {
-    this.eventUpdates[event.id] = event;
+    if (this.bufferedEvents.has(event.id)) {
+      const buffered = this.bufferedEvents.get(event.id)!;
+      if (event === buffered) return;
+      else Object.assign(buffered, event);
+    } else this.eventUpdates[event.id] = event;
   }
 
   abandon(): void {
-    if (this.stream?.close()) rmSync(this.partPath);
-    this.stream = undefined;
+    if (this.running && this.stream?.close()) rmSync(this.partPath);
+    this.running = false;
+    this.disposeBufferedEvents(Recording.rootBuffer);
   }
 
   finish(): boolean {
+    if (!this.running) return false;
+    this.passEvents(this.stream, Recording.rootBuffer);
     const written = this.stream?.close(
       compactObject({
         classMap: makeClassMap(this.functionsSeen.keys()),
         metadata: compactObject(this.metadata),
         eventUpdates: Object.keys(this.eventUpdates).length > 0 ? this.eventUpdates : undefined,
       }),
     );
-    this.stream = undefined;
     if (written) {
       renameSync(this.partPath, this.path);
       writtenAppMaps.push(this.path);
     }
+    this.running = false;
+    this.disposeBufferedEvents(Recording.rootBuffer);
     return !!written;
   }
 
-  get running(): boolean {
-    return !!this.stream;
+  public running = true;
+
+  private bufferedEvents = new Map<number, AppMap.Event>();
+
+  public emit(event: AppMap.Event) {
+    if (!this.running) {
+      warn("event emitted while recording not running");
+      return;
+    }
+    // If the current buffer is alive more than allowed pass its events
+    // to the stream and clear it recursively.
+    if (Recording.buffering && !Recording.buffer.disposed && config.asyncTrackingTimeout != 0) {
+      const elapsed = getTimeInMilliseconds() - Recording.buffer.createdAt;
+      if (elapsed >= config.asyncTrackingTimeout)
```

Comment:
Perhaps just use `getTime()` to avoid having a duplicate function that does the same thing? And rename `config.asyncTrackingTimeout` to `config.asyncTrackingTimeoutMs` to avoid confusion.
```suggestion
      const elapsed = getTime() - Recording.buffer.createdAt;
      if (1000 * elapsed >= config.asyncTrackingTimeoutMs)
```

Reply:
Actually, let's leave it in as is. `hrtime()` seems deprecated.